src/transactions/TransactionUtils.cpp

```diff
@@ -1142,34 +1142,245 @@ claimableBalanceFlagIsValid(ClaimableBalanceEntry const& cb)
         }
+        ltxInner.prefetch(poolKeys);
 
-        releaseLiabilities(ltxInner, header, offer);
-        auto trustAcc = stellar::loadAccount(ltxInner, account);
-        removeEntryWithPossibleSponsorship(ltxInner, header, offer.current(),
-                                           trustAcc);
-        offer.erase();
+        for (auto& poolShareTrustLine : poolShareTrustLines)
+        {
+            // use a lambda so we don't hold a reference to the internals of
+            // TrustLineEntry
+            auto poolTL = [&]() -> TrustLineEntry const& {
+                return poolShareTrustLine.current().data.trustLine();
+            };
+
+            auto pool =
+                loadLiquidityPool(ltxInner, poolTL().asset.liquidityPoolID());
+            auto balance = poolTL().balance;
+            auto cbSponsoringAccID =
+                poolTL().ext.v() == 1 &&
+                        getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                            .sponsoringID
+                    ? *getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                           .sponsoringID
+                    : accountID;
+
+            // release reserves for pool share trustline that will be deleted
+            // later
+            {
+                auto trustAcc = stellar::loadAccount(ltxInner, accountID);
+                removeEntryWithPossibleSponsorship(
+                    ltxInner, ltxInner.loadHeader(),
+                    poolShareTrustLine.current(), trustAcc);
+            }
+
+            auto redeemIntoClaimableBalance =
+                [&ltxInner, &txSourceID, &txSeqNum, &opIndex, &poolTL,
+                 &accountID, &cbSponsoringAccID](Asset const& assetInPool,
+                                                 int64_t amount) -> bool {
+                // create a claimable balance for the asset in the pool
+                LedgerEntry claimableBalance;
+                claimableBalance.data.type(CLAIMABLE_BALANCE);
+
+                auto& claimableBalanceEntry =
+                    claimableBalance.data.claimableBalance();
+
+                OperationID operationID;
+                operationID.type(ENVELOPE_TYPE_POOL_REVOKE_OP_ID);
+                operationID.revokeId().sourceAccount = txSourceID;
+                operationID.revokeId().seqNum = txSeqNum;
+                operationID.revokeId().opNum = opIndex;
+                operationID.revokeId().liquidityPoolID =
+                    poolTL().asset.liquidityPoolID();
+                operationID.revokeId().asset = assetInPool;
+
+                claimableBalanceEntry.balanceID.type(
+                    CLAIMABLE_BALANCE_ID_TYPE_FROM_POOL_REVOKE);
+                claimableBalanceEntry.balanceID.fromPoolRevoke() =
+                    xdrSha256(operationID);
+
+                claimableBalanceEntry.amount = amount;
+                claimableBalanceEntry.asset = assetInPool;
+
+                ClaimPredicate u;
```

Comment:
You might consider refactoring this and the next two blocks into a `makeRevokeClaimants` function (or something similar).

Reply:
Nitpick: also for this use case, I don't see a lot of reason to copy the `ClaimPredicate` in. You could have just written `c.v0().predicate.type(CLAIM_PREDICATE_UNCONDITIONAL)` on line 1265.

Reply:
Updated to use `makeUnconditionalClaimant`.